apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/SetOptions.cs

```diff
@@ -0,0 +1,83 @@
+    /// Immutable class representing options for Set operations.
+    /// </summary>
+    public sealed class SetOptions
+    {
+        /// <summary>
+        /// Returns an instance that overwrites the target object. This is the default when no
+        /// options are provided.
+        /// </summary>
+        public static SetOptions Overwrite { get; } = new SetOptions(false, new FieldPath[0]);
+
+        /// <summary>
+        /// Returns an instance that merges all fields.
+        /// </summary>
+        public static SetOptions MergeAll { get; } = new SetOptions(true, new FieldPath[0]);
+
+        /// <summary>
+        /// Returns an instance that merges the given fields.
+        /// </summary>
+        /// <param name="fieldMask">The fields to merge. An empty array is equivalent to using <see cref="MergeAll"/>.
+        /// Must not be null or contain any empty or null elements. Each field is treated as a dot-separated list of segments.
+        /// </param>
+        /// <returns>An instance that merges the given fields.</returns>
+        public static SetOptions MergeFields(params string[] fieldMask)
+        {
+            GaxPreconditions.CheckNotNull(fieldMask, nameof(fieldMask));
+            GaxPreconditions.CheckArgument(fieldMask.All(x => !string.IsNullOrEmpty(x)), nameof(fieldMask), "Field mask must not contain any null or empty elements");
+            return new SetOptions(true, fieldMask.Select(field => FieldPath.FromDotSeparatedString(field)).ToArray());
+        }
+
+        /// <summary>
+        /// Returns an instance that merges the given fields.
+        /// </summary>
+        /// <param name="fieldMask">The fields to merge. An empty array is equivalent to using <see cref="MergeAll"/>.
+        /// Must not be null or contain any null elements.</param>
+        /// <returns>An instance that merges the given fields.</returns>
+        public static SetOptions MergeFields(params FieldPath[] fieldMask)
+        {
+            GaxPreconditions.CheckNotNull(fieldMask, nameof(fieldMask));
+            GaxPreconditions.CheckArgument(fieldMask.All(x => x != null), nameof(fieldMask), "Field mask must not contain any null elements");
+            // Note: field mask is copied for the sake of immutability.
+            return new SetOptions(true, fieldMask.ToArray());
+        }
+
+        private SetOptions(bool merge, FieldPath[] fieldMask)
+        {
+            Merge = merge;
+            FieldMask = new ReadOnlyCollection<FieldPath>(fieldMask);
+        }
+
+        /// <summary>
+        /// Whether or not to merge values. If <see cref="FieldMask"/> is non-empty,
+        /// this is always <c>true</c>.
+        /// </summary>
+        public bool Merge { get; }
+
+        /// <summary>
+        /// The fields to merge. This property is never null; if the mask is empty,
+        /// all fields will be merged or overwritten.
+        /// </summary>
+        public ReadOnlyCollection<FieldPath> FieldMask { get; }
```

Comment:
Would `IReadOnlyCollection` be better?

Reply:
Hmm. Or maybe `IReadOnlyList`... will check that works with netstandard1.5 and go to that.